cl/opinion_page/forms.py

```diff
@@ -53,6 +54,19 @@ class CitationRedirectorForm(forms.Form):
         required=False,
     )
 
+    def clean_reporter(self):
+        data = self.cleaned_data["reporter"]
+        # Flag reporter text for potential URLs
+        if bool(re.search(r"http[s]?://\S+", data)):
+            raise ValidationError(
+                (
+                    "URLs are not allowed in this field. "
+                    "Please remove the URL and try again."
+                )
+            )
+        # Sanitize reporter text: remove slashes
+        return data.replace("/", " ")
```

Comment:
I'm surprised we need to do this. Can't eyecite just ignore bad input and move on? Like, what if my input is:

```
The https://  protocol was discussed in 22 U.S. 44
```

Shouldn't that work just fine? I thought eyecite would be really good at dealing with that?

Reply:
> Can't eyecite just ignore bad input and move on?

Yeah Eyecite can handle the input, but we're adding a sanitization step for the reporter field because that string is not directly passed to the `get_citations` method. The `citation_homepage` view is the one that handles the form submission and the view redirects users to another view(`citation_redirector`) where the actual analysis of the input value happens.

If the value of the 'reporter' field contains forward slashes ('/'), the redirection logic in the 'citation_homepage' view fails because none of the URL patterns used for the 'citation_redirector' view allow slashes.

while writing this explanation, I realized there's a simpler solution😞. Instead of sanitizing the input text in the clean method, we can update the `citation_homepage` view to use the `get_citations` method from Eyecite. This way, the view will only redirect to the `citation_redirector` view when the input contains citations. By doing this, we can rely on Eyecite to help us sanitize the input text and we can remove `clean_reporter` method.

Reply:
Here's an update about this comment:

I've tried using Eyecite alone to handle the form submission within the `citation_homepage` view. While this approach works well for most scenarios, there's a specific limitation. The current implementation of the tool allows users to search by reporter name only. For instance, if a user enters "U.S." in the search bar, we should display all volumes associated with the **United States Supreme Court Reporter**. Unfortunately, Eyecite is not able to identify the reporter's name in such cases and returns an empty list.

if we still want to allow users to search by reporter name only, we need to send the text they enter to the `citation_redirection` view. However, we must ensure the text is sanitized and doesn't contain slashes. Therefore, we should keep the `clean_reporter` method around.

Reply:
Does it get simpler if we stop supporting search for just the reporter?

Reply:
> Does it get simpler if we stop supporting search for just the reporter?

Code-wise, If we ditch the reporter search feature, we can axe the `clean_reporter` method too (it's only like 3 lines of code anyway!) and that's pretty much it.

OTOH, supporting the reporter search feature makes the logic to show a "no citation found" error as you suggested in [here](https://github.com/freelawproject/courtlistener/pull/3916#discussion_r1544050430) a little bit confusing because an empty list from Eyecite isn't enough to display the message(The input might be a reporter name). We would only be able to show it when Eyecite finds citations within the text and after filtering the list, we find that none of them is a case law citation 🤔 

Here's a table summarizing the behavior of the new `no-citation-found` error with and without reporter search support:

| Input | Eyecite Ouput  | Reporter Search On |  Reporter Search Off
| ------------- | ------------- | ------------- | ------------- |
| a bc (Random text)  | Empty list  |  bad-reporter-error | no-citation-found(new error) |
| U.S. (Reporter name)  | Empty list  | List of volumes | no-citation-found(new error) |
|  Iowa Code § 602.1614. | List with a LawCitation |  no-citation-found(new error) | no-citation-found(new error) |
|  §102 USC 222, 44 Vand. L. Rev. 1041 (1991) | List with a LawCitation and a JournalCitation | no-citation-found(new error) | no-citation-found(new error) |
| 613 N.W.2d 120  | List with a CaseCitation |  Clusters or no-match-error | Clusters or no-match-error |

Reply:
I think we just need to refine the error messages and we can still support reporter search:

| Input | Eyecite Ouput  | Response |  
| ------------- | ------------- | ------------- | 
| a bc (Random text)  | Empty list  |  **No reporter or citation error(replaces the bad-reporter error)** | 
| U.S. (Reporter name)  | Empty list  | List of volumes | 
|  Iowa Code § 602.1614. | List with a LawCitation |  **no-caselaw-citation-found(new error)** | 
|  §102 USC 222, 44 Vand. L. Rev. 1041 (1991) | List with a LawCitation and a JournalCitation | **no-caselaw-citation-found(new error)** | 
| 613 N.W.2d 120  | List with a CaseCitation |  Opinion page, List of Clusters or no-match-error |


- By refining the "bad-reporter" error, we can provide the user with clearer feedback. The new message can indicate that the input either doesn't contain any valid citations or a recognizable reporter name.

- The new error message you suggested [here](https://github.com/freelawproject/courtlistener/pull/3916#discussion_r1544050430) could inform users that their input might contain citations, but not for legal opinions. We can refine the existing *Citation Type Mismatch* error introduced by this PR.

@mlissner let me know what you think.

Reply:
I think looking up just reporters really isn't very useful, and it adds complexity to the system, so I don't think it's worth it even if you could make it work. I'm also thinking that the cleaning step feels wack-a-mole. Like, we can take care of slashes and `https`, but eventually we'll find some other funky thing we have to deal with.

Reply:
Alright! I'll remove the code to stop supporting reporter searches via the search bar. However, we can still access all volumes associated with a reporter using the URL. Thanks for your input! 👍 🚀

Reply:
Perfect. Thank you!

Reply:
Here's a GIF showing how the citation search works with the new logic:

![Screen Recording 2024-03-29 at 5 19 08 PM](https://github.com/freelawproject/courtlistener/assets/55959657/919188a4-f2bb-4db2-b138-653898d5c826)


I tested the following scenarios:

- Random Text 
- Reporter name
- Non-caselaw citations:
   - Iowa Code § 602.1614
   - §102 USC 222, 44 Vand. L. Rev. 1041 (1991)
- Caselaw citation: 
   - 613 N.W.2d 12
- Listing the volumes using the URL
- A citation with no volume